src/Http/Routing/src/Builder/RoutingEndpointConventionBuilderExtensions.cs

```diff
@@ -40,6 +40,57 @@ public static class RoutingEndpointConventionBuilderExtensions
             return builder;
         }
 
+        /// <summary>
+        /// Requires that endpoints match the specified header during routing.
+        /// </summary>
+        /// <param name="builder">The <see cref="IEndpointConventionBuilder"/> to add the metadata to.</param>
+        /// <param name="headerName">Header name to match.</param>
+        /// <param name="headerValues">
+        /// Acceptable header values that should match.
+        /// An empty collection means any value will be accepted as long as the header is present.
+        /// </param>
+        /// <returns>A reference to this instance after the operation has completed.</returns>
+        public static TBuilder RequireHeader<TBuilder>(this TBuilder builder, string headerName, params string[] headerValues) where TBuilder : IEndpointConventionBuilder
+        {
+            return RequireHeader(builder, headerName, HeaderValueMatchMode.Exact, false, headerValues);
+        }
+
+        /// <summary>
+        /// Requires that endpoints match the specified header during routing.
+        /// </summary>
+        /// <param name="builder">The <see cref="IEndpointConventionBuilder"/> to add the metadata to.</param>
+        /// <param name="headerName">Header name to match.</param>
+        /// <param name="valueMatchMode">Specifies how header values should be compared (e.g. exact matches Vs. by prefix).</param>
+        /// <param name="valueIgnoresCase">Specifies whether header value comparisons should ignore case.</param>
+        /// <param name="headerValues">
+        /// Acceptable header values that should match.
+        /// An empty collection means any value will be accepted as long as the header is present.
+        /// </param>
+        /// <returns>A reference to this instance after the operation has completed.</returns>
+        public static TBuilder RequireHeader<TBuilder>(this TBuilder builder, string headerName, HeaderValueMatchMode valueMatchMode, bool valueIgnoresCase, params string[] headerValues) where TBuilder : IEndpointConventionBuilder
+        {
+            if (headerName == null)
+            {
+                throw new ArgumentNullException(nameof(headerName));
+            }
+
+            if (headerValues == null)
+            {
+                throw new ArgumentNullException(nameof(headerValues));
+            }
+
+            builder.Add(endpointBuilder =>
+            {
+                endpointBuilder.Metadata.Add(
+                    new HeaderAttribute(headerName, headerValues)
```

Comment:
We normally have different types for the attribute and the metadata, for example for HTTP methods we have HttpMethodAttribute and HttpMethodMetadata. That gives us more flexibility to decide how we want to interpret the attributes at higher levels of the stack like MVC and similar.